// wdio-interface/build/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _sourceMapSupport2 = require('source-map-support');

var _util = require('util');

var _util2 = _interopRequireDefault(_util);

var _ansiEscapes = require('ansi-escapes');

var _ansiEscapes2 = _interopRequireDefault(_ansiEscapes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

(0, _sourceMapSupport2.install)();
class CLInterface {
    constructor() {
        var _context;

        this.i = 0;
        this.stdoutBuffer = [];
        this.stderrBuffer = [];
        this.out = (_context = process.stdout).write.bind(_context);
        this.err = (_context = process.stderr).write.bind(_context);

        this.clearAll();

        /**
         * don't modify stdout and stderr streams for unit tests
         */
        /* istanbul ignore if */
        if (!process.env.WDIO_TEST) {
            this.wrapStdio(process.stdout, this.stdoutBuffer);
            this.wrapStdio(process.stderr, this.stderrBuffer);
        }
    }

    wrapStdio(stream, buffer) {
        stream.write = chunk => {
            buffer.push(chunk);
            return true;
        };
    }

    clearAll() {
        this.out(_ansiEscapes2.default.clearScreen);
    }

    clearLine() {
        this.out(_ansiEscapes2.default.eraseStartLine);
        this.out(_ansiEscapes2.default.cursorLeft);
    }

    log(...messages) {
        this.out(_util2.default.format.apply(this, messages) + '\n');
    }

    write(message) {
        this.out(message);
    }
}
exports.default = CLInterface;
module.exports = exports['default'];
