// webdriverio/build/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.multiremote = exports.attach = exports.remote = void 0;

require("source-map-support/register");

var _webdriver = _interopRequireDefault(require("webdriver"));

var _config = require("@wdio/config");

var _multiremote = _interopRequireDefault(require("./multiremote"));

var _constants = require("./constants");

var _utils = require("./utils");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * A method to create a new session with WebdriverIO
 *
 * @param  {Object} [params={}]       Options to create the session with
 * @param  {function} remoteModifier  Modifier function to change the monad object
 * @return {object}                   browser object with sessionId
 */
const remote = function (params = {}, remoteModifier) {
  const config = (0, _config.validateConfig)(_constants.WDIO_DEFAULTS, params);

  const modifier = (client, options) => {
    if (typeof remoteModifier === 'function') {
      client = remoteModifier(client, Object.assign(options, config));
    }

    Object.assign(options, config);
    return client;
  };

  if (params.user && params.key) {
    params = Object.assign(params, (0, _config.detectBackend)(params));
  }

  const prototype = (0, _utils.getPrototype)('browser');
  /**
   * apply mobile check flags to browser scope
   */

  const mobileDetection = (0, _utils.mobileDetector)(params.capabilities);
  Object.assign(prototype, Object.keys(mobileDetection).reduce((proto, flag) => {
    proto[flag] = {
      value: mobileDetection[flag]
    };
    return proto;
  }, {}));
  return _webdriver.default.newSession(params, modifier, prototype, _config.wrapCommand);
};

exports.remote = remote;

const attach = function (params) {
  const prototype = (0, _utils.getPrototype)('browser');
  return _webdriver.default.attachToSession(params, null, prototype, _config.wrapCommand);
};

exports.attach = attach;

const multiremote = async function (params = {}) {
  const multibrowser = new _multiremote.default();
  const browserNames = Object.keys(params);
  /**
   * create all instance sessions
   */

  await Promise.all(browserNames.map(browserName => {
    (0, _config.validateConfig)(_constants.WDIO_DEFAULTS, params[browserName]);
    const prototype = (0, _utils.getPrototype)('browser');

    const instance = _webdriver.default.newSession(params[browserName], null, prototype);

    return multibrowser.addInstance(browserName, instance);
  }));
  /**
   * use attachToSession capability to wrap instances around blank pod
   */

  const prototype = (0, _utils.getPrototype)('browser');
  const sessionParams = {
    sessionId: '',
    isW3C: multibrowser.instances[browserNames[0]].isW3C
  };
  return _webdriver.default.attachToSession(sessionParams, multibrowser.modifier.bind(multibrowser), prototype);
};

exports.multiremote = multiremote;
