// wdio-cli/build/interface.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _sourceMapSupport2 = require('source-map-support');

var _slicedToArray = function () { function sliceIterator(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"]) _i["return"](); } finally { if (_d) throw _e; } } return _arr; } return function (arr, i) { if (Array.isArray(arr)) { return arr; } else if (Symbol.iterator in Object(arr)) { return sliceIterator(arr, i); } else { throw new TypeError("Invalid attempt to destructure non-iterable instance"); } }; }();

var _chalk = require('chalk');

var _chalk2 = _interopRequireDefault(_chalk);

var _cliSpinners = require('cli-spinners');

var _cliSpinners2 = _interopRequireDefault(_cliSpinners);

var _events = require('events');

var _events2 = _interopRequireDefault(_events);

var _wdioInterface = require('wdio-interface');

var _wdioInterface2 = _interopRequireDefault(_wdioInterface);

var _wdioLogger = require('wdio-logger');

var _wdioLogger2 = _interopRequireDefault(_wdioLogger);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

(0, _sourceMapSupport2.install)();


const log = (0, _wdioLogger2.default)('wdio-cli');

const clockSpinner = _cliSpinners2.default['clock'];

class WDIOCLInterface extends _events2.default {
    constructor(config, specs) {
        super();
        this.hasAnsiSupport = !!_chalk2.default.supportsColor.hasBasic;
        this.clockTimer = 0;
        this.specs = specs;
        this.config = config;
        this.jobs = new Map();
        this.start = Date.now();
        this.result = {
            finished: 0,
            passed: 0,
            failed: 0
        };
        this.messages = {
            reporter: {}
        };

        this.interface = new _wdioInterface2.default();
        this.on('job:start', this.addJob.bind(this));
        this.on('job:end', this.clearJob.bind(this));
    }

    /**
     * add job to interface
     */
    addJob({ cid, caps, specs }) {
        this.jobs.set(cid, { caps, specs });
        this.updateView();
    }

    /**
     * clear job from interface
     */
    clearJob({ cid, passed }) {
        this.jobs.delete(cid);
        this.result.finished++;

        if (passed) {
            this.result.passed++;
        } else {
            this.result.failed++;
        }

        this.updateView(true);
    }

    /**
     * event handler that is triggered when runner sends up events
     */
    onMessage(params) {
        if (!params.origin || !this.messages[params.origin]) {
            return log.warn(`Can't identify message from worker: ${JSON.stringify(params)}, ignoring!`);
        }

        if (!this.messages[params.origin][params.name]) {
            this.messages[params.origin][params.name] = [];
        }
        this.messages[params.origin][params.name].push(params.content);
    }

    updateView(wasJobCleared) {
        const isFinished = this.jobs.size === 0;
        const pendingJobs = this.specs.length - this.jobs.size - this.result.finished;

        /**
         * check if environment supports ansi and print a limited update if not
         */
        if (!this.hasAnsiSupport && !isFinished) {
            /**
             * only update if a job finishes
             */
            if (!wasJobCleared) {
                return;
            }

            const clockSpinnerSymbol = this.getClockSymbol();
            return this.interface.log(`${clockSpinnerSymbol} ` + `${this.jobs.size} running, ` + `${this.result.passed} passed, ` + `${this.result.failed} failed, ` + `${this.specs.length} total ` + `(${Math.round(this.result.finished / this.specs.length * 100)}% completed)`);
        }

        this.interface.clearAll();
        this.interface.log();

        /**
         * print reporter output
         */
        var _iteratorNormalCompletion = true;
        var _didIteratorError = false;
        var _iteratorError = undefined;

        try {
            for (var _iterator = Object.entries(this.messages.reporter)[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
                const _ref = _step.value;

                var _ref2 = _slicedToArray(_ref, 2);

                const reporterName = _ref2[0];
                const messages = _ref2[1];

                this.interface.log(_chalk2.default.bgYellow.black(`"${reporterName}" Reporter:`));
                this.interface.log(messages.join(''));
                this.interface.log();
            }

            /**
             * print running jobs
             */
        } catch (err) {
            _didIteratorError = true;
            _iteratorError = err;
        } finally {
            try {
                if (!_iteratorNormalCompletion && _iterator.return) {
                    _iterator.return();
                }
            } finally {
                if (_didIteratorError) {
                    throw _iteratorError;
                }
            }
        }

        var _iteratorNormalCompletion2 = true;
        var _didIteratorError2 = false;
        var _iteratorError2 = undefined;

        try {
            for (var _iterator2 = this.jobs.entries()[Symbol.iterator](), _step2; !(_iteratorNormalCompletion2 = (_step2 = _iterator2.next()).done); _iteratorNormalCompletion2 = true) {
                const _ref3 = _step2.value;

                var _ref4 = _slicedToArray(_ref3, 2);

                const cid = _ref4[0];
                const job = _ref4[1];

                const filename = job.specs.join(', ').replace(process.cwd(), '');
                this.interface.log(_chalk2.default.bgYellow.black(' RUNNING '), cid, 'in', job.caps.browserName, '-', filename);
            }
        } catch (err) {
            _didIteratorError2 = true;
            _iteratorError2 = err;
        } finally {
            try {
                if (!_iteratorNormalCompletion2 && _iterator2.return) {
                    _iterator2.return();
                }
            } finally {
                if (_didIteratorError2) {
                    throw _iteratorError2;
                }
            }
        }

        if (pendingJobs) {
            this.interface.log(_chalk2.default.yellow('...', pendingJobs, 'pending tests'));
        }

        /**
         * print stdout and stderr from runners
         */
        if (isFinished) {
            if (this.interface.stdoutBuffer.length) {
                this.interface.log(_chalk2.default.bgYellow.black(`Stdout:\n`) + this.interface.stdoutBuffer.join(''));
            }
            if (this.interface.stderrBuffer.length) {
                this.interface.log(_chalk2.default.bgRed.black(`Stderr:\n`) + this.interface.stderrBuffer.join(''));
            }
        }

        /**
         * add empty line between "pending tests" and results
         */
        if (this.jobs.size) {
            this.interface.log();
        }

        this.interface.log('Test Suites:\t', _chalk2.default.green(this.result.passed, 'passed') + ', ' + (this.result.failed ? _chalk2.default.red(this.result.failed, 'failed') + ', ' : '') + this.specs.length, 'total', `(${this.specs.length ? Math.round(this.result.finished / this.specs.length * 100) : 0}% completed)`);

        this.updateClock();

        if (isFinished) {
            clearTimeout(this.interval);
            this.interface.log('\n');
        }
    }

    updateClock(interval = 100) {
        const clockSpinnerSymbol = this.getClockSymbol();

        clearTimeout(this.interval);
        this.interface.clearLine();
        this.interface.write('Time:\t\t ' + clockSpinnerSymbol + ' ' + ((Date.now() - this.start) / 1000).toFixed(2) + 's');
        this.interval = setTimeout(() => this.updateClock(interval), interval);
    }

    getClockSymbol() {
        return clockSpinner.frames[this.clockTimer = ++this.clockTimer % clockSpinner.frames.length];
    }
}
exports.default = WDIOCLInterface;
module.exports = exports['default'];
