// wdio-logger/build/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _sourceMapSupport2 = require('source-map-support');

exports.default = getLogger;

var _fs = require('fs');

var _fs2 = _interopRequireDefault(_fs);

var _loglevel = require('loglevel');

var _loglevel2 = _interopRequireDefault(_loglevel);

var _util = require('util');

var _util2 = _interopRequireDefault(_util);

var _chalk = require('chalk');

var _chalk2 = _interopRequireDefault(_chalk);

var _loglevelPluginPrefix = require('loglevel-plugin-prefix');

var _loglevelPluginPrefix2 = _interopRequireDefault(_loglevelPluginPrefix);

var _stripAnsi = require('strip-ansi');

var _stripAnsi2 = _interopRequireDefault(_stripAnsi);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

(0, _sourceMapSupport2.install)();


const DEFAULT_LEVEL = 0;
const COLORS = {
    error: 'red',
    warn: 'yellow',
    info: 'cyanBright',
    debug: 'green',
    trace: 'cyan'
};

const SERIALIZERS = [{
    /**
     * display error stack
     */
    matches: err => err instanceof Error,
    serialize: err => err.stack
}, {
    /**
     * color commands blue
     */
    matches: log => log === 'COMMAND',
    serialize: log => _chalk2.default.magenta(log)
}, {
    /**
     * color data yellow
     */
    matches: log => log === 'DATA',
    serialize: log => _chalk2.default.yellow(log)
}, {
    /**
     * color result cyan
     */
    matches: log => log === 'RESULT',
    serialize: log => _chalk2.default.cyan(log)
}];

const loggers = {};
const logCache = new Set();
let logFile;

const originalFactory = _loglevel2.default.methodFactory;
_loglevel2.default.methodFactory = function (methodName, logLevel, loggerName) {
    const rawMethod = originalFactory(methodName, logLevel, loggerName);
    return (...args) => {
        /**
         * create logFile lazily
         */
        if (!logFile && process.env.WDIO_LOG_PATH) {
            logFile = _fs2.default.createWriteStream(process.env.WDIO_LOG_PATH);
        }

        args = args.map(arg => {
            var _iteratorNormalCompletion = true;
            var _didIteratorError = false;
            var _iteratorError = undefined;

            try {
                for (var _iterator = SERIALIZERS[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
                    const s = _step.value;

                    if (s.matches(arg)) {
                        return s.serialize(arg);
                    }
                }
            } catch (err) {
                _didIteratorError = true;
                _iteratorError = err;
            } finally {
                try {
                    if (!_iteratorNormalCompletion && _iterator.return) {
                        _iterator.return();
                    }
                } finally {
                    if (_didIteratorError) {
                        throw _iteratorError;
                    }
                }
            }

            return arg;
        });

        const logText = (0, _stripAnsi2.default)(`${_util2.default.format.apply(this, args)}\n`);
        if (logFile) {
            /**
             * empty logging cache if stuff got logged before
             */
            if (logCache.size) {
                logCache.forEach(log => logFile.write(log));
                logCache.clear();
            }

            return logFile.write(logText);
        }

        logCache.add(logText);
        rawMethod(...args);
    };
};

_loglevelPluginPrefix2.default.apply(_loglevel2.default, {
    template: '%t %l %n:',
    timestampFormatter: date => _chalk2.default.gray(date.toISOString()),
    levelFormatter: level => _chalk2.default[COLORS[level]](level.toUpperCase()),
    nameFormatter: name => _chalk2.default.whiteBright(name || 'global')
});

function getLogger(name) {
    /**
     * check if logger was already initiated
     */
    if (loggers[name]) {
        return loggers[name];
    }

    loggers[name] = _loglevel2.default.getLogger(name);
    loggers[name].setLevel(process.env.WDIO_LOG_LEVEL || DEFAULT_LEVEL);
    return loggers[name];
}

getLogger.setLevel = (name, level) => loggers[name].setLevel(level);
module.exports = exports['default'];
