// wdio-config/build/shim.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.hasWdioSyncSupport = exports.wrapCommand = exports.runFnInFiberContext = exports.runTestInFiberContext = exports.executeHooksWithArgs = undefined;

var _sourceMapSupport2 = require('source-map-support');

var _wdioLogger = require('wdio-logger');

var _wdioLogger2 = _interopRequireDefault(_wdioLogger);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _asyncToGenerator(fn) { return function () { var gen = fn.apply(this, arguments); return new Promise(function (resolve, reject) { function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { return Promise.resolve(value).then(function (value) { step("next", value); }, function (err) { step("throw", err); }); } } return step("next"); }); }; }

(0, _sourceMapSupport2.install)();


const log = (0, _wdioLogger2.default)('wdio-config');
const NOOP = () => {};

let executeHooksWithArgs = exports.executeHooksWithArgs = (() => {
    var _ref = _asyncToGenerator(function* (hooks, args) {
        /**
         * make sure hooks are an array of functions
         */
        if (!Array.isArray(hooks)) {
            hooks = [hooks];
        }

        /**
         * make sure args is an array since we are calling apply
         */
        if (!Array.isArray(args)) {
            args = [args];
        }

        hooks = hooks.map(function (hook) {
            return new Promise(function (resolve) {
                let result;

                try {
                    result = hook.apply(null, args);
                } catch (e) {
                    log.error(e.stack);
                    return resolve();
                }

                /**
                 * if a promise is returned make sure we don't have a catch handler
                 * so in case of a rejection it won't cause the hook to fail
                 */
                if (result && typeof result.then === 'function') {
                    return result.then(resolve, function (e) {
                        log.error(e.stack);
                        resolve();
                    });
                }

                resolve(result);
            });
        });

        return Promise.all(hooks);
    });

    function executeHooksWithArgsShim(_x, _x2) {
        return _ref.apply(this, arguments);
    }

    return executeHooksWithArgsShim;
})();

let runTestInFiberContext = exports.runTestInFiberContext = NOOP;
let runFnInFiberContext = exports.runFnInFiberContext = (fn, done) => {
    return function (...args) {
        const result = fn.apply(this, args);

        if (typeof done === 'function') {
            return done(result);
        }

        return result;
    };
};
let wrapCommand = exports.wrapCommand = (_, origFn) => origFn;
let hasWdioSyncSupport = exports.hasWdioSyncSupport = false;

/**
 * shim to make sure that we only wrap commands if wdio-sync is installed as dependency
 */
try {
    // eslint-disable-next-line import/no-unresolved
    const wdioSync = require('wdio-sync');
    log.debug('wdio-sync found, running tests synchronous');

    exports.hasWdioSyncSupport = hasWdioSyncSupport = true;
    exports.runFnInFiberContext = runFnInFiberContext = wdioSync.runFnInFiberContext;
    exports.runTestInFiberContext = runTestInFiberContext = wdioSync.runTestInFiberContext;
    exports.wrapCommand = wrapCommand = wdioSync.wrapCommand;
    exports.executeHooksWithArgs = executeHooksWithArgs = wdioSync.executeHooksWithArgs;
} catch (e) {
    log.debug('wdio-sync not found, running tests asynchronous');
}
